[Web API][Bluetooth] Update API reference and guides (#1284)
Samsung/tizen-docs: docs/application/web/guides/connectivity/bluetooth.md

You can use Bluetooth functionalities in your application, such as managing the local Bluetooth adapter, bonding, and exchanging data between Bluetooth-enabled devices. The Bluetooth standard provides a peer-to-peer (P2P) data exchange functionality over short distance between compliant devices.

> ## Prerequisites

Author: I have applied your suggestions.

> To use the Application (in [mobile](../../api/latest/device_api/mobile/tizen/application.html), [wearable](../../api/latest/device_api/wearable/tizen/application.html) and [tv](../../api/latest/device_api/tv/tizen/application.html applications) and Bluetooth (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html) applications) APIs, the application has to request permission by adding the following privileges to the `config.xml` file:

Reviewer: Global: Please verify all the links

> To use the Application (in [mobile](../../api/latest/device_api/mobile/tizen/application.html), [wearable](../../api/latest/device_api/wearable/tizen/application.html) and [tv](../../api/latest/device_api/tv/tizen/application.html applications) and Bluetooth (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html) applications) APIs, the application has to request permission by adding the following privileges to the `config.xml` file:

Author: I have applied your suggestions.

> To use the Application (in [mobile](../../api/latest/device_api/mobile/tizen/application.html), [wearable](../../api/latest/device_api/wearable/tizen/application.html) and [tv](../../api/latest/device_api/tv/tizen/application.html applications) and Bluetooth (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html) applications) APIs, the application has to request permission by adding the following privileges to the `config.xml` file:

Author: Done

```
<tizen:privilege name="http://tizen.org/privilege/application.launch"/>
<tizen:privilege name="http://tizen.org/privilege/bluetooth"/>
```

> ## Handling UUIDs and binary data in Bluetooth API

Author: I have applied your suggestions.

> ### Handling UUIDs

Author: I have applied your suggestions.

> According to the [Bluetooth Core Specification](https://www.bluetooth.com/specifications/bluetooth-core-specification/), UUIDs used to represent Bluetooth objects can take 3 forms:

Author: I have applied your suggestions.

   * 128-bit representation: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX", where each 'X' stands for a hexadecimal digit.  
   For example: "198d3a9c-e21a-4f72-a48b-39a6bad7e583".
   * 32-bit representation: "D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>D<sub>5</sub>D<sub>6</sub>D<sub>7</sub>D<sub>8</sub>", equivalent to "D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>D<sub>5</sub>D<sub>6</sub>D<sub>7</sub>D<sub>8</sub>-0000-1000-8000-00805F9B34FB", where D<sub>1</sub>..D<sub>8</sub> stand for hexadecimal digits.  
   For example: "e72ad71b".
   * 16-bit representation: "D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>", equivalent to "0000D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>-0000-1000-8000-00805F9B34FB", where D<sub>1</sub>..D<sub>4</sub> stand for hexadecimal digits.  
   For example: "d182".

Unless stated otherwise, each function taking a UUID as a parameter will accept any of the 3 forms and make appropriate conversions.  
Unless stated otherwise, each UUID returned from a function and each UUID attribute of an object may be a 16-bit, 32-bit or 128-bit UUID.

> Bluetooth API functions are case-insensitive with regard to UUIDs - lowercase (abcdef) and uppercase (ABCDEF) characters are recognized as valid hexadecimal digits.

Author: > Request you to please rephrase to follow Tizen writing style.

Could you please give me a link to the Tizen writing style? I have tried to find it in this repository, but have only found the [markdown style guide ](https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md).
 
> Request you to please rephrase to follow Tizen writing style. Sentences like "Unless statd otherwise" is not recommended. For example,
> 
> Each function taking a UUID as a parameter accepts any of the three forms and makes appropriate conversions.
> Each UUID returned from a function and each UUID attribute of an object can be a 16-bit, 32-bit, or 128-bit UUID.

I am afraid, that rephrasing these sentences in this way would cause a loss of some, quite important, part of their meaning. We used the "Unless stated otherwise, ..." part to tell the reader, that some exceptions to these general rules are possible and they are stated in the documentation.

What do you think about adding the following sentence after them, to clarify, that there may be some exceptions to this rule?

> All exceptions to these rules are noted in the API reference for [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html) devices.

> Bluetooth API functions are case-insensitive with regard to UUIDs - lowercase (abcdef) and uppercase (ABCDEF) characters are recognized as valid hexadecimal digits.

Author: I have applied the rest of your suggestions

The 128-bit UUID that is the base for UUIDs having 16-bit or 32-bit equivalents is defined available through the API in `BluetoothManager` as `BASE_UUID`.

`BluetoothManager` methods to manipulate UUIDs:

> 

Author: Done

      uuidFrom16bits is equal to "00001234-0000-1000-8000-00805f9b34fb"  
      uuidFrom32bits is equal to "ab5690ef-0000-1000-8000-00805f9b34fb"  
      uuidFrom128bits is equal to "abcdef01-2345-6789-abcd-ef0123456789"

   * `uuidToShortestPossible()`
      ```
      var from16Bit = tizen.bluetooth.uuidToShortestPossible("1234");
      var from32Bit = tizen.bluetooth.uuidToShortestPossible("0000acdf");
      var from128BitFirst = tizen.bluetooth.uuidToShortestPossible("ab5690ef-0000-1000-8000-00805F9B34FB");
      var from128BitSecond = tizen.bluetooth.uuidToShortestPossible("abcdef01-2345-6789-abcd-ef0123456789");
      ```

      from16Bit is equal to "1234"  
      from32Bit is equal to "acdf"  
      from128BitFirst is equal to "ab5690ef"  
      from128BitSecond is equal to "abcdef01-2345-6789-abcd-ef0123456789"

   * `uuidsEqual()`
      ```
      var first = tizen.bluetooth.uuidsEqual("1234", "00001234");
      var second = tizen.bluetooth.uuidsEqual("ab5690ef", "ab5690ef-0000-1000-8000-00805F9B34FB");
      var third = tizen.bluetooth.uuidsEqual("abcdef01-2345-6789-abcd-ef0123456789", "abcdef01");
      ```

      Both `first` and `second` are `true`. `third` is false.

> ### Handling binary data

Author: I have applied your suggestions.

> The `Bytes` type, that aggregates all types in Bluetooth API used to pass binary data. It can be either a *byte[]* or a *DOMString* or a *Uint8Array*.

Author: Done

`BluetoothManager` methods to manipulate binary data:

>    * `toByteArray()`

Author: Done

      var dataString = "0x18b1";

      var first = tizen.bluetooth.toByteArray(dataInt8Array);
      var second = tizen.bluetooth.toByteArray(dataUint8Array);
      var third = tizen.bluetooth.toByteArray(dataString);
      ```
      `first`, `second` and `third` variables are equal arrays.

   * `toDOMString()`
      ```
      var dataInt8Array = new Int8Array([24, 177]);
      var dataUint8Array = new Uint8Array([24, 177]);
      var dataString = "0x18b1";

      var first = tizen.bluetooth.toDOMString(dataInt8Array);
      var second = tizen.bluetooth.toDOMString(dataUint8Array);
      var third = tizen.bluetooth.toDOMString(dataString);
      ```
      `first`, `second` and `third` variables are equal arrays.

   * `toUint8Array()`
      ```
      var dataInt8Array = new Int8Array([24, 177]);
      var dataUint8Array = new Uint8Array([24, 177]);
      var dataString = "0x18b1";

      var first = tizen.bluetooth.toUint8Array(dataInt8Array);
      var second = tizen.bluetooth.toUint8Array(dataUint8Array);
      var third = tizen.bluetooth.toUint8Array(dataString);
      ```
      `first`, `second` and `third` variables are equal.

> ## The main features of the Bluetooth API include:

Author: I have removed the colon from the heading and kept the original titles.

- Managing the local Bluetooth adapter

  You can [enable and disable the local Bluetooth adapter](#managing-the-local-bluetooth-adapter), and change the device name for it.

- Discovering devices

  You can [discover other Bluetooth devices](#discovering-bluetooth-devices).

- Creating a bond with a Bluetooth device

  You can [create a bond](#creating-a-bond-with-a-bluetooth-device) with another device retrieved through the discovery process. Bonding allows the 2 devices to establish a connection.

- Connecting to and exchanging data with a Bluetooth device

  You can [connect to and exchange data with a remote Bluetooth device](#connecting-to-and-exchanging-data-with-a-bluetooth-device).

> ## The main Bluetooth (4.0) Low Energy features include:

Reviewer: Please avoid using colons in heading. Suggestion

## Main features of Bluetooth (4.0) Low Energy

The main features of the  Bluetooth (4.0) Low Energy features include:

Author: I have applied your suggestions.

- Retrieving Bluetooth GATT services

  You can [retrieve information about Bluetooth GATT services](#retrieving-bluetooth-gatt-services) provided by the remote device.

  Every GATT service defines characteristics it includes. By knowing the service, you know what features the Bluetooth device exposes.

- Accessing the Bluetooth GATT characteristic value

  You can [read and write the Bluetooth GATT characteristic value](#accessing-the-bluetooth-gatt-characteristic-value).

  Characteristics allows you to monitor and sometimes control remote Bluetooth Low Energy devices. For example, a sensor reading can be exposed by the sensor device as a Bluetooth GATT characteristic.

- Receiving notifications on characteristic value changes

  You can [monitor a characteristic value](#receiving-notifications-on-characteristic-value-changes) to detect any changes, for example, in sensor readings and battery level.

- Accessing the Bluetooth GATT descriptor value

  You can [read and write the Bluetooth GATT descriptor value](#accessing-the-bluetooth-gatt-descriptor-value).

## Managing the Local Bluetooth Adapter

> You can enable or disable the local Bluetooth adapter, and set the device name using the system-provided service through the `ApplicationControl` interface (in [mobile](../../api/latest/device_api/mobile/tizen/application.html#ApplicationControl), [wearable](../../api/latest/device_api/wearable/tizen/application.html#ApplicationControl) and [tv](../../api/latest/device_api/tv/tizen/application.html#ApplicationControl) applications).

Reviewer: Global: Please change all headings to verb form and sentence case

Author: I have applied your suggestions.

To use the Bluetooth functionality of the device, you must switch the Bluetooth adapter on. The Bluetooth API does not provide a method to enable or disable the Bluetooth adapter of the device directly. When Bluetooth is required, you must request the built-in Settings application on the device to let the user enable or disable Bluetooth.

**Figure: Bluetooth setting screen**

![Bluetooth setting screen](./media/bluetooth_onoff.png)

> 1. Retrieve a `BluetoothAdapter` object (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothAdapter), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothAdapter) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothAdapter) applications) with the `getDefaultAdapter()` method and prepare the `ApplicationControl` object (in [mobile](../../api/latest/device_api/mobile/tizen/application.html#ApplicationControl), [wearable](../../api/latest/device_api/wearable/tizen/application.html#ApplicationControl) and [tv](../../api/latest/device_api/tv/tizen/application.html#ApplicationControl) applications) to request the Bluetooth switching operation:

Reviewer: Please add a serial comma before tv

Author: I have added serial commas before "tv" in the whole file.

You can use Bluetooth functionalities in your application, such as managing the local Bluetooth adapter, bonding, and exchanging data between Bluetooth-enabled devices. The Bluetooth standard provides a peer-to-peer (P2P) data exchange functionality over short distance between compliant devices.

## Main features of Bluetooth API

The main features of the Bluetooth API include:

- Handle UUIDs and binary data

  You can [convert UUIDs and binary data between equivalent formats and compare UUIDs in different formats](#handle-uuids-and-binary-data).

- Manage the local Bluetooth adapter

  You can [enable and disable the local Bluetooth adapter](#manage-the-local-bluetooth-adapter), and change the device name for it.

- Discover devices

  You can [discover other Bluetooth devices](#discover-bluetooth-devices).

- Create a bond with a Bluetooth device

>   You can [create a bond](#create-a-bond-with-a-bluetooth-device) with another device retrieved through the discovery process. Bonding allows the 2 devices to establish a connection.

Reviewer: Please change 2 devices to two devices:

  You can [create a bond](#create-a-bond-with-a-bluetooth-device) with another device retrieved through the discovery process. Bonding allows the two devices to establish a connection.

Author: Done

- Receive notifications on characteristic value changes

  You can [monitor a characteristic value](#receive-notifications-on-characteristic-value-changes) to detect any changes, for example, in sensor readings and battery level.

- Access the Bluetooth GATT descriptor value

  You can [read and write the Bluetooth GATT descriptor value](#access-the-bluetooth-gatt-descriptor-value).

- Monitor ATT MTU of a connection with a remote GATT server

  You can [read the current value of connection's ATT MTU](#access-the-att-mtu-of-connected-device) and [register a listener to receive information about its changes](#receive-notifications-on-att-mtu-changes).

- Manage the advertise options

  You can [manage advertise](#manage-the-advertise-options) to control how your device announces itself to other Bluetooth Low Energy devices for discovery.

- Start and stop the local GATT server

  You can [start](#start-the-server) and [stop](#stop-the-server) exposing services registered in the local GATT server.

- Receive notifications on GATT connection state changes

>   You can [register a listener to monitor GATT connection changes](#receive-notifications-on-connection-state-changes).

Author: Yes, indeed.

- Enable reading and writing values of characteristics and descriptors registered in the local GATT server

  You can [register callbacks to respond to read and write value requests from server clients](#respond-to-read-and-write-value-requests-from-server-clients).

## Prerequisites

To use the Application (in [mobile](../../api/latest/device_api/mobile/tizen/application.html), [wearable](../../api/latest/device_api/wearable/tizen/application.html), and [tv](../../api/latest/device_api/tv/tizen/application.html) applications) and Bluetooth (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html) applications) APIs, the application has to request permission by adding the following privileges to the `config.xml` file:

```
<tizen:privilege name="http://tizen.org/privilege/application.launch"/>
<tizen:privilege name="http://tizen.org/privilege/bluetooth"/>
```
## Handle UUIDs and binary data

With Bluetooth API, you can handle UUIDs and binary data.

### Handle UUIDs

> According to the [Bluetooth Core Specification](https://www.bluetooth.com/specifications/bluetooth-core-specification/), UUIDs that are used to represent Bluetooth objects can take the following 3 forms:

Reviewer: Please change 3 to three:

According to the [Bluetooth Core Specification](https://www.bluetooth.com/specifications/bluetooth-core-specification/), UUIDs that are used to represent Bluetooth objects can take the following three forms:

Author: Done

   * 128-bit representation: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX", where each 'X' stands for a hexadecimal digit.  
   For example: "198d3a9c-e21a-4f72-a48b-39a6bad7e583".
   * 32-bit representation: "D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>D<sub>5</sub>D<sub>6</sub>D<sub>7</sub>D<sub>8</sub>", equivalent to "D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>D<sub>5</sub>D<sub>6</sub>D<sub>7</sub>D<sub>8</sub>-0000-1000-8000-00805F9B34FB", where D<sub>1</sub>..D<sub>8</sub> stand for hexadecimal digits.  
   For example: "e72ad71b".

>    * 16-bit representation: "D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>", equivalent to "0000D<sub>1</sub>D<sub>2</sub>D<sub>3</sub>D<sub>4</sub>-0000-1000-8000-00805F9B34FB", where D<sub>1</sub>..D<sub>4</sub> stand for hexadecimal digits.  

Author: Done

      var from32Bit = tizen.bluetooth.uuidToShortestPossible("0000acdf");
      var from128BitFirst = tizen.bluetooth.uuidToShortestPossible("ab5690ef-0000-1000-8000-00805F9B34FB");
      var from128BitSecond = tizen.bluetooth.uuidToShortestPossible("abcdef01-2345-6789-abcd-ef0123456789");
      ```

      from16Bit is equal to "1234"  
      from32Bit is equal to "acdf"  
      from128BitFirst is equal to "ab5690ef"  
      from128BitSecond is equal to "abcdef01-2345-6789-abcd-ef0123456789"

   - `uuidsEqual()` tests if two UUIDs have a single common equivalent UUID
      ```
      var first = tizen.bluetooth.uuidsEqual("1234", "00001234");
      var second = tizen.bluetooth.uuidsEqual("ab5690ef", "ab5690ef-0000-1000-8000-00805F9B34FB");
      var third = tizen.bluetooth.uuidsEqual("abcdef01-2345-6789-abcd-ef0123456789", "abcdef01");
      ```

      Both `first` and `second` are `true`. `third` is false.

### Handle binary data
The `Bytes` type, that aggregates all types in Bluetooth API used to pass binary data. It can be either a `byte[]` or a `DOMString` or a `Uint8Array`.

`BluetoothManager` methods to manipulate binary data:

>    - `toByteArray()` converts a variable from any of the `Bytes` types to `byte[]`

Reviewer: Please add a colon:

   - `toByteArray()` converts a variable from any of the `Bytes` types to `byte[]`:

Author: Done

      ```
      var dataInt8Array = new Int8Array([24, 177]);
      var dataUint8Array = new Uint8Array([24, 177]);
      var dataString = "0x18b1";

      var first = tizen.bluetooth.toByteArray(dataInt8Array);
      var second = tizen.bluetooth.toByteArray(dataUint8Array);
      var third = tizen.bluetooth.toByteArray(dataString);
      ```

>       `first`, `second` and `third` variables are equal arrays.

Reviewer: Missing serial comma:

      `first`, `second`, and `third` variables are equal arrays.

Author: Done

>    - `toDOMString()` converts a variable from any of the `Bytes` types to `DOMString`

Reviewer: Please add a colon:

   - `toDOMString()` converts a variable from any of the `Bytes` types to `DOMString`:

Author: Done

      ```
      var dataInt8Array = new Int8Array([24, 177]);
      var dataUint8Array = new Uint8Array([24, 177]);
      var dataString = "0x18b1";

      var first = tizen.bluetooth.toDOMString(dataInt8Array);
      var second = tizen.bluetooth.toDOMString(dataUint8Array);
      var third = tizen.bluetooth.toDOMString(dataString);
      ```
      `first`, `second` and `third` variables are equal arrays.

>    - `toUint8Array()` converts a variable from any of the `Bytes` types to `Uint8Array`

Reviewer: Please add a colon:

   - `toUint8Array()` converts a variable from any of the `Bytes` types to `Uint8Array`:

Author: Done

   adapter.setName(chatServerName);
   ```

## Managing remote devices

> ### Discovering Bluetooth Devices

Reviewer: ## Manage remote devices

### Discover Bluetooth Devices

Author: I have applied your suggestions.

   tizen.application.launchAppControl(bluetoothVisibilityAppControl, null, null, launchVisibilityError, serviceVisibilityReply);
   ```

> 6. Set a friendly name for the device using the `setName()` method. The name helps to recognize the device in a list of [retrieved devices](#discover-bluetooth-devices).

Reviewer: For step 5 as well as step 6, please use a colon instead of a period as it is followed by a ciode:

6. Set a friendly name for the device using the `setName()` method. The name helps to recognize the device in a list of [retrieved devices](#discover-bluetooth-devices):

Author: Done

You can also retrieve the known devices which were bonded or found in a prior discovery process.

To search for remote devices and get the known devices:

1. Retrieve a `BluetoothAdapter` object (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothAdapter), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothAdapter), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothAdapter) applications) with the `getDefaultAdapter()` method:

   ```
   var adapter = tizen.bluetooth.getDefaultAdapter();
   ```

2. To search for remote devices, use the `discoverDevices()` method.

>    The results of the search are returned in the `BluetoothDiscoverDevicesSuccessCallback` (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothDiscoverDevicesSuccessCallback), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothDiscoverDevicesSuccessCallback), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothDiscoverDevicesSuccessCallback) applications).

Reviewer: Please use a colon instead of a period as it is followed by a code:

   The results of the search are returned in the `BluetoothDiscoverDevicesSuccessCallback` (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothDiscoverDevicesSuccessCallback), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothDiscoverDevicesSuccessCallback), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothDiscoverDevicesSuccessCallback) applications):

Author: Done

6. Set a friendly name for the device using the `setName()` method.

>    The name helps to recognize the device in a list of [retrieved devices](#discover-bluetooth-devices).

Reviewer: Could we combine this sentence with the previous sentence:

6. Set a friendly name for the device using the `setName()` method. The name helps to recognize the device in a list of [retrieved devices](#discover-bluetooth-devices):

Author: Done

   adapter.discoverDevices(discoverDevicesSuccessCallback, null);
   ```

   > [!NOTE]
   > To allow other Bluetooth devices to find your device, you must set the device to be visible through the system settings.

3. To retrieve known devices (which have been previously paired or searched for), use the `getKnownDevices()` method.

>    The results of the search are returned in the `BluetoothDeviceArraySuccessCallback` (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothDeviceArraySuccessCallback), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothDeviceArraySuccessCallback), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothDeviceArraySuccessCallback) applications).

Reviewer: Please use a colon instead of a period as it is followed by a code:

   The results of the search are returned in the `BluetoothDeviceArraySuccessCallback` (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothDeviceArraySuccessCallback), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothDeviceArraySuccessCallback), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothDeviceArraySuccessCallback) applications):

Author: Done

   adapter.createBonding('35:F4:59:D1:7A:03', onBondingSuccessCallback, onErrorCallback);
   ```

>    > **Note**

Reviewer: Global: Please use the latest note tag:

https://github.com/Samsung/tizen-docs/blob/master/styleguide/custom-style.md

Author: I have applied your suggestions.

   > The MAC address of the Bluetooth device is a `BluetoothAddress` object (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothAddress), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothAddress) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothAddress) applications). You can get the MAC address of the peer device from the `BluetoothDevice` object (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothDevice), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothDevice) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothDevice) applications), which is returned in the success callback of the `BluetoothAdapter`'s `getKnownDevices()` and `discoverDevices()` methods.

3. To end the bond with a remote device, use the `destroyBonding()` method:

   ```
   adapter.destroyBonding('35:F4:59:D1:7A:03');
   ```

> ### Connecting to and Exchanging Data with a Bluetooth Device

Reviewer: ### Connect to and exchange data with Bluetooth device

Author: I have applied your suggestions.

The Radio Frequency Communication (RFCOMM) is a set of transport protocols which allows multiple simultaneous connections to a device. If a device allows other devices to use its functionalities through this kind of connection, it is said to provide a service and it is called a server device. The devices that request the service are called client devices.

To connect to services provided by a server device to the client devices:

1. Retrieve a `BluetoothAdapter` object (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothAdapter), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothAdapter), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothAdapter) applications) with the `getDefaultAdapter()` method:

>    `var adapter = tizen.bluetooth.getDefaultAdapter();`

Author: Done

   When an incoming message is received from the peer device, the `onmessage` event handler in the `BluetoothSocket` interface is triggered.

> ### Discovering Bluetooth Low Energy Devices

Reviewer: ### Discover Bluetooth Low Energy devices

Author: I have applied your suggestions.

   descriptor.writeValue(newValue, writeSuccess, writeFail);
   ```

> ### Accessing the ATT MTU of connected device

Reviewer: Suggestion:

### Access ATT MTU of connected device

Author: I have applied your suggestions.

> To get the ATT MTU value or request change of the ATT MTU value:

Reviewer: Suggestion:
To get the ATT MTU value or request change of the ATT MTU value, follow these steps:

Author: I have applied your suggestions.

1. [Connect to a Bluetooth Low Energy device](#connecting-to-a-bluetooth-low-energy-device).

> 2. If the `device` has been connected call `getAttMtu()` function on the `device` object to get current ATT MTU value:

Reviewer: Global: Please remove the word function:

2. If the `device` is connected, call `getAttMtu()` on the `device` object to get the current ATT MTU value:

Author: I have applied your suggestions.

   ```
   var attMtu = device.getAttMtu();
   ```

> 3. If the `device` has been connected call `requestAttMtuChange()` function on the `device` object to request a change of the ATT MTU value. Pass desired ATT MTU value as an argument:

Reviewer: 3. If the `device` is connected, call `requestAttMtuChange()` on the `device` object to request a change of the ATT MTU value. Pass desired ATT MTU value as an argument:

Author: I have applied your suggestions.

   ```
   var newAttMtuValue = 64;
   device.requestAttMtuChange(newAttMtuValue);
   ```

>    > **Note**

Reviewer: Please use the update note tag

Author: I have applied your suggestions.

   > After calling `requestAttMtuChange()` function ATT MTU value change should be accepted if both devices support new ATT MTU value according to the [Bluetooth Core Specification](https://www.bluetooth.com/specifications/bluetooth-core-specification/).

### Receiving Notifications on ATT MTU changes

To recieve notifications on ATT MTU value changes:

1. [Connect to a Bluetooth Low Energy device](#connecting-to-a-bluetooth-low-energy-device).

> 2. If the `device` has been connected call `addAttMtuChangeListener()` function providing as a parameter the callback function to be called on each change of the ATT MTU value:

Author: Done

   ```
   function attMtuChangeCallback(newAttMtuValue) {
      console.log("ATT MTU value has changed to: " + newAttMtuValue);
   }
   var listenerId = device.addAttMtuChangeListener(attMtuChangeCallback);
   ```
   The `listenerId` value stores an identifier of the listener, which is needed to remove the listener.
3. After setting up the listener with `addAttMtuChangeListener()` function, the callback can be invoked by changing the ATT MTU value:
   ```
   var newAttMtuValue = 50;
   device.requestAttMtuChange(newAttMtuValue);
   ```
   The change of ATT MTU value will trigger the callback function.
   > **Note**
   > After calling `requestAttMtuChange()` function ATT MTU value change should be accepted if both devices support new ATT MTU value according to the [Bluetooth Core Specification](https://www.bluetooth.com/specifications/bluetooth-core-specification/).

> When a listener monitoring the ATT MTU value changes is no longer needed, you can remove it:

Author: Done

4. Call `removeAttMtuChangeListener()` providing the identifier of the listener you want to remove.
   ```
   device.removeAttMtuChangeListener(listenerId);
   ```
   After removing of the listener, changes of the ATT MTU value will no longer trigger the callback function.

## Managing the local GATT server

### Start the server

To start the local GATT server:

1. Retrieve `BluetoothGATTServer` object (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothGATTServer), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothGATTServer) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothGATTServer) applications) with `getGATTServer()` method of the `BluetoothManager` interface (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothManager), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothManager) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothManager) applications):

```
var server = tizen.bluetooth.getGATTServer();

> ```

Author: Done

2. Set up callbacks and start the local GATT server with `start()` method of `BluetoothGATTServer` (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothGATTServer), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothGATTServer) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothGATTServer) applications) interface:

```
var onSuccess = function() {
   console.log("Server started successfully");
};

var onError = function(error) {
   console.error("Failed to start the server, error: " + error.message);
};

server.start(onSuccess, onError);

> ```

Author: Done

### Stop the server

To stop the local GATT server:

1. Retrieve `BluetoothGATTServer` object (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothGATTServer), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothGATTServer) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothGATTServer) applications) with `getGATTServer()` method of the `BluetoothManager` interface (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothManager), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothManager) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothManager) applications):

```
var server = tizen.bluetooth.getGATTServer();

> ```

Author: Done

2. Set up callbacks and stop the local GATT server with `stop()` method of `BluetoothGATTServer` (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothGATTServer), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothGATTServer) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothGATTServer) applications) interface:

```
var onSuccess = function() {
   console.log("Server stopped successfully");
};

var onError = function(error) {
   console.error("Failed to stop the server, error: " + error.message);
};

server.stop(onSuccess, onError);

> ```

Author: Done

> ### Receiving notifications on server's connection state changes

Author: Done

To receive notifications whenever a remote connection with the local server is established or lost:

1. Define `BluetoothLEConnectChangeCallback`:

```
var connectChangeCallback = {
   onconnected: function(device) {
      console.log("A client connected: " + device.address);
   },
   ondisconnected: function(device) {
      console.log("A client disconnected: " + device.address);
   }
};

> ```

Author: Done

2. Add the defined callback with `addConnectStateChangeListener()` method (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothLEAdapter::addConnectStateChangeListener), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothLEAdapter::addConnectStateChangeListener) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothLEAdapter::addConnectStateChangeListener) applications):

```
var adapter = tizen.bluetooth.getLEAdapter();
var watchId = adapter.addConnectStateChangeListener(connectChangeCallback);

> ```

Author: Done

> The `watchId` value stores an identifier of the listener, which is needed to remove the listener.

Author: Done

3. When a listener monitoring the connection changes is no longer needed, it can be removed:

```
adapter.removeConnectStateChangeListener(watchId);

> ```

Author: Done

### Accessing connection's ATT MTU

To get the value of connection's ATT MTU value:

1. Define the `ConnectionMtuCallback` and (optionally) the error callback:

```
var connectionMtuCB = function(mtu) {
   console.log("The ATT MTU value: " + mtu);
};

var errorCB = function(error) {
   console.log("Cannot retrieve ATT MTU value, error: " + error.message);
};
```

2. Select proper remote client's address and get the value of ATT MTU with `getConnectionMtu()` method (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothGATTServer::getConnectionMtu), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothGATTServer::getConnectionMtu) and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothGATTServer::getConnectionMtu) applications):

```
var clientAddress = "12:34:56:78:90:ab";
var server = tizen.bluetooth.getGATTServer();
server.getConnectionMtu(clientAddress, connectionMtuCB, errorCB);

> ```

Author: Done

   };

   var server = tizen.bluetooth.getGATTServer();

   server.registerService(gattService, registerServiceSuccesCB, registerServiceErrorCB);
   ```

3. Start the GATT server:
   ```
   var serverStartSuccessCB = function() {
      console.log("Server started successfully!");
   };

   var serverStartErrorCB = function(error) {
      console.error("Server didn't start, error: " + error.name + "; " + error.message);
   };

   server.start(serverStartSuccessCB, serverStartErrorCB);
   ```

4. Optional callbacks that are used by the functions that set the callbacks to the read value requests and the write value requests should be defined:
   ```
   var setCallbackSuccessCB() {
      console.log("Callback set successfully!");
   };

   var setCallbackErrorCB = function(error) {
      console.error("Callback is not set correctly, error: " + error.name + "; " + error.message);
   };

   var sendResponseSuccessCB() {
      console.log("Response sent successfully!");
   };

   var sendResponseErrorCB = function(error) {
      console.error("Response send failure, error: " + error.name + "; " + error.message);
   };
   ```

> 5. To register the callback called when a client reads the value of the characteristic from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback():

Author: Done

   ```
   var characteristicWriteRequestCallback = function(clientAddress, value, offset, replyRequired) {
      console.log(clientAddress + " requested to write characteristic's value: " + value +
                  " with offset: " + offset);
      return replyRequired ? new tizen.GATTRequestReply(0) : null;
   };

   server.services[0].characteristics[0].setWriteValueRequestCallback(characteristicWriteRequestCallback, setCallbackSuccessCB, setCallbackErrorCB, sendResponseSuccessCB, sendResponseErrorCB);
   ```

   > **Note**
   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this characteristic.

7. To register the callback called when a client reads the value of the descriptor from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the descriptor object.
   ```
   var descriptorReadRequestCallback = function(clientAddress, offset) {
      console.log(clientAddress + " requested to read descriptor's value with offset: " + offset);
      return new tizen.GATTRequestReply(0, "0x1234");
   };

>    server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(characteristicReadRequestCallback, setCallbackSuccessCB, setCallbackErrorCB, sendResponseSuccessCB, sendResponseErrorCB);

Reviewer 2: Question. According to the line 1218, it seems the descriptorReadRequestCallback of line 1220 should be the argument, but characteristicReadRequestCallback is. Is it ok?

Author: Done

   ```
   var adapter = tizen.bluetooth.getLEAdapter();
   ```

2. Set up options and start advertising with the `startAdvertise()` method of the `BluetoothLEAdapter` interface:

   ```
   var advertiseData = new tizen.BluetoothLEAdvertiseData({
       includeName: true,
       serviceuuids: ['180f'] /* 180F is 16bit Battery Service UUID */
   });
   var connectable = true;

   adapter.startAdvertise(advertiseData, 'ADVERTISE', function onstate(state) {
       console.log('Advertising configured: ' + state);
   }, function(error) {
       console.log('startAdvertise() failed: ' + error.message);
   }, 'LOW_LATENCY', connectable);
   ```

   > **Note**
   > To learn how to make your mobile device visible to other Bluetooth devices, see [Managing the Local Bluetooth Adapter](#managing-the-local-bluetooth-adapter).

3. To disable the advertising, use the `stopAdvertise()` method of the `BluetoothLEAdapter` interface:

   ```
   adapter.stopAdvertise();
   ```

## Related information

> * Dependencies

Author: Done

   - Tizen 2.4 and Higher for Mobile

>    - Tizen 2.3.1 and Higher for Wearable

Author: Done

   descriptor.writeValue(newValue, writeSuccess, writeFail);
   ```

### Access the ATT MTU of connected device

To get the ATT MTU value or request change of the ATT MTU value, follow these steps:

1. [Connect to a Bluetooth Low Energy device](#connect-to-a-bluetooth-low-energy-device).

> 2. If the `device` has been connected call `getAttMtu()` on the `device` object to get current ATT MTU value:

Author: I have now looked for all "function" occurrences in the file and removed all the redundant ones.

      console.error("Failed to stop the server, error: " + error.message);
   };

   server.stop(onSuccess, onError);
   ```

### Receive notifications on GATT connection state changes

To receive notifications whenever a GATT connection with the device is established or lost:

1. Define `BluetoothLEConnectChangeCallback`:

   ```
   var connectChangeCallback = {
      onconnected: function(device) {
         console.log("A device connected: " + device.address);
      },
      ondisconnected: function(device) {
         console.log("A device disconnected: " + device.address);
      }
   };
   ```

2. Add the defined callback with `addConnectStateChangeListener()` method (in [mobile](../../api/latest/device_api/mobile/tizen/bluetooth.html#BluetoothLEAdapter::addConnectStateChangeListener), [wearable](../../api/latest/device_api/wearable/tizen/bluetooth.html#BluetoothLEAdapter::addConnectStateChangeListener), and [tv](../../api/latest/device_api/tv/tizen/bluetooth.html#BluetoothLEAdapter::addConnectStateChangeListener) applications):

   ```
   var adapter = tizen.bluetooth.getLEAdapter();
   var watchId = adapter.addConnectStateChangeListener(connectChangeCallback);
   ```

> The `watchId` value stores an identifier of the listener, which is needed to remove the listener.

Author: Done

   }

   serviceToBeUnregistered.unregister(successCallback, errorCallback);
   ```
   > [!NOTE]
   > After `unregister()` is called for the last registered service, the server is stopped.
   > If new services are then registered, server's `start()` method has to be
   > called to make them visible to clients.

To unregister all services from the local GATT server at once, call server's
`unregisterAllServices()` method:

   ```
   function successCallback() {
     console.log('All services unregistered successfully');
   }

   function errorCallback(error) {
     console.log('Unregistering all GATT services failed: ' + error.message);
   }

   var server = tizen.bluetooth.getGATTServer();
   server.unregisterAllServices(successCallback, errorCallback);
   ```
   > [!NOTE]
   > After the calling `unregisterAllServices()`, the server is stopped.
   > If new services are then registered, server's `start()` method has to be
   > called to make them visible to clients.

### Send notifications about characteristic's value changes to the clients
GATT clients connected to the server running on the local device can register for updates of its characteristics' values.

> The server can send two types of such updates - _notifications_ and _indications_.

Reviewer 2: In the Tizen docs, we don't recommend to use italic font. So please change " _notifications_ and _indications_" to " `notifications` and `indications`".

The server can send two types of such updates, `notifications` and `indications`.

@surbhig12 Would you please also check the sentence?

Reviewer: The sentence is fine.

Author: Done

> They differ in that, clients receiving indications have to acknowledge them - i.e. send a message back to the server, telling that they received the new value. Clients do not acknowledge notifications.

Reviewer 2: @surbhig12 Would you please check this sentence again?

They differ in that, clients receiving indications have to acknowledge them - i.e. send a message back to the server, telling that they received the new value. Clients do not acknowledge notifications.

Reviewer: Please let me know if the followign sentence is technically fine:

The difference in notifications and indications is that clients receiving indications have to acknowledge them. This means that the client must send a message back to the server, telling that they received the new value. Clients do not acknowledge notifications.

Please verify if the last sentence is correct: Clients do not acknowledge notifications.

Please verify for technical accuracy.

Author: @surbhig12, The sentences you have proposed are technically correct.

I have one question about the second sentence:
> This means that the client must send a message back to the server, telling that they received the new value.

Shouldn't clients be plural here, like:
> This means that the _clients_ must send a message back to the server, telling that they received the new value.
?

For now, I change the sentence to the version with plural clients. If you have meant singular `the client`, please let me know and I'll fix it.

Reviewer: Thank you for the clarification, the rephrased sentence is fine.

Notifications and indications are not enabled in characteristics by default.
To enable notifications or indications in a characteristic:

1. Set its `isNotify` or `isIndication` property:

   ```
   var notificationEnabledCharacteristicInitData = {
     uuid: '1234',
     isNotify: true // or set isIndication to enable indications
     /* Define more attributes, if needed */
   };
   ```

> 2. The characteristic requires a special kind of descriptor, called _Client Characteristic Configuration Descriptor_ (CCCD).
> For more details about CCCD, refer to the _Bluetooth Core Specification_, available at [bluetooth.com](https://bluetooth.com).

Reviewer 2: @surbhig12 Would you please check again these sentences?

```suggestion
2. The characteristic requires a special kind of descriptor, Client Characteristic Configuration Descriptor (CCCD).
For more details about CCCD, see [Bluetooth Core Specification](https://www.bluetooth.com/specifications/bluetooth-core-specification/).
```
@pwasowski2 I hope you to check the updated link also.

Author: The link is OK. I have applied your suggestions.

   };

   var server = tizen.bluetooth.getGATTServer();

   server.registerService(gattService, registerServiceSuccesCB, registerServiceErrorCB);
   ```

3. Start the GATT server:
   ```
   var serverStartSuccessCB = function() {
      console.log("Server started successfully!");
   };

   var serverStartErrorCB = function(error) {
      console.error("Server didn't start, error: " + error.name + "; " + error.message);
   };

   server.start(serverStartSuccessCB, serverStartErrorCB);
   ```

4. Optional callbacks that are used by the functions that set the callbacks to the read value requests and the write value requests should be defined:
   ```
   var setCallbackSuccessCB() {
      console.log("Callback set successfully!");
   };

   var setCallbackErrorCB = function(error) {
      console.error("Callback is not set correctly, error: " + error.name + "; " + error.message);
   };

   var sendResponseSuccessCB() {
      console.log("Response sent successfully!");
   };

   var sendResponseErrorCB = function(error) {
      console.error("Response send failure, error: " + error.name + "; " + error.message);
   };
   ```

> 5. To register the callback called when a client reads the value of the characteristic from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()`:

Reviewer: Please remove article "the":

5. To register the callback called when a client reads the value of the characteristic from the local GATT server, create the callback function and pass it as an argument to `setReadValueRequestCallback()`:

Author: Done
